src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8938,74 +8894,77 @@ private ContinueStatementSyntax ParseContinueStatement(SyntaxList<AttributeListS
-                    catches = _pool.Allocate<CatchClauseSyntax>();
+                    catchClauses = _pool.Allocate<CatchClauseSyntax>();
                     while (this.CurrentToken.Kind == SyntaxKind.CatchKeyword)
-                    {
-                        catches.Add(this.ParseCatchClause());
-                    }
+                        catchClauses.Add(this.ParseCatchClause());
                 }
 
                 if (this.CurrentToken.Kind == SyntaxKind.FinallyKeyword)
                 {
-                    hasEnd = true;
-                    var fin = this.EatToken();
-                    var finBlock = this.ParsePossiblyAttributedBlock();
-                    @finally = _syntaxFactory.FinallyClause(fin, finBlock);
+                    finallyClause = _syntaxFactory.FinallyClause(
+                        this.EatToken(),
+                        this.ParsePossiblyAttributedBlock());
                 }
 
-                if (!hasEnd)
+                if (catchClauses.IsNull && finallyClause == null)
                 {
-                    block = this.AddErrorToLastToken(block, ErrorCode.ERR_ExpectedEndTry);
+                    if (!ContainsErrorDiagnostic(tryBlock))
+                        tryBlock = this.AddErrorToLastToken(tryBlock, ErrorCode.ERR_ExpectedEndTry);
 
                     // synthesize missing tokens for "finally { }":
-                    @finally = _syntaxFactory.FinallyClause(
-                        SyntaxToken.CreateMissing(SyntaxKind.FinallyKeyword, null, null),
-                        _syntaxFactory.Block(
-                            attributeLists: default,
-                            SyntaxToken.CreateMissing(SyntaxKind.OpenBraceToken, null, null),
-                            default(SyntaxList<StatementSyntax>),
-                            SyntaxToken.CreateMissing(SyntaxKind.CloseBraceToken, null, null)));
+                    finallyClause = _syntaxFactory.FinallyClause(
+                        createMissing(SyntaxKind.FinallyKeyword),
+                        missingBlock());
                 }
 
-                _isInTry = isInTry;
-
-                return _syntaxFactory.TryStatement(attributes, @try, block, catches, @finally);
+                return _syntaxFactory.TryStatement(attributes, @try, tryBlock, catchClauses, finallyClause);
             }
             finally
             {
-                if (!catches.IsNull)
-                {
-                    _pool.Free(catches);
-                }
+                if (!catchClauses.IsNull)
+                    _pool.Free(catchClauses);
             }
+
+            BlockSyntax missingBlock()
+                => _syntaxFactory.Block(
+                    attributeLists: default,
+                    createMissing(SyntaxKind.OpenBraceToken),
+                    statements: default,
+                    createMissing(SyntaxKind.CloseBraceToken));
+
+            static SyntaxToken createMissing(SyntaxKind kind)
```

Comment:
This local function doesn't feel necessary. Other code does: `SyntaxFactory.MissingToken(SyntaxKind.XYZ)`

Reply:
totally true.  i missed that!